### PR TITLE
Replace AsyncWrite with Upload trait and rename MultiPartStore to MultipartStore (#5458)

### LINKED ISSUE #5443: In Object Store, return version & etag on multipart put.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

To protect integrity of our uploads, we store object versions together with s3 file paths. This is currently easy for regular `ObjectStore::put` requests, but is not possible with multipart uploads when using `ObjectStore::put_multipart`. The current return values of the latter (MultipartId and AsyncWrite) do not allow for an easy return of values, so maybe it's possible to add this to `WriteMultiPart` (so that we don't reimplement it using just `MultiPartStore`/`PutPart` traits).

**Describe the solution you'd like**

1. (Optional) Make it possible to construct a PutPart from a MultiPartStore. Currently when you do `create_multipart`, it just returns a multipart id. Maybe add a `create_multipart_put_part(&self, path: &Path) -> Result<Box<dyn PutPart>>`? This will enable using WriteMultiPart on any MultiPartStore from the client code. This is good because then users can tune WriteMultiPart parameters like concurrency and potentially part size.
1. `PutPart::complete` should return `Result<PutResult>`
2. `WriteMultiPart` should remember this PutResult and provide a method to take it after shutdown is polled to completion.

After a bit more analysis, it looks like PutPart and MultiPartStore are duplicating put_part and complete methods. Plus, it looks like you can create a generic PutPart implementation given just the MultiPartStore interface. What do you think about simplifying this? There are 2 options:
 1. Only keep create_multipart method in MultiPartStore, make it return PutPart. Add "abort" and "get_upload_id" methods to the PutPart. This way 
 2. Keep MultiPartStore as-is, remove PutPart trait completely, make WriteMultiPart work with MultiPartStore directly. This will require adding location and upload_id to WriteMultiPart.

**Describe alternatives you've considered**

* Reimplement WriteMultiPart logic with the changes above. There's a lot of code there though.
* Issue a HEAD request right after the put. This will work in most cases, but is not guaranteed.

**Additional context**
I'm using version 0.9.

### LINKED ISSUE #5458: Revisit Design of ObjectStore::put_multipart
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently streaming uploads are supported by `ObjectStore::put_multipart`. This returns a `AsyncWrite`, which provides a push-based interface for writing data.

However, this approach is not without issue:

* No obvious way to return PutResult for parts or the final object - https://github.com/apache/arrow-rs/issues/5443
* No obvious way to retry uploading of a single part - https://github.com/apache/arrow-rs/issues/5437
* Unclear poisoning behaviour - https://github.com/apache/arrow-rs/issues/5437
* Cannot support resuming uploads - https://github.com/apache/arrow-rs/issues/4961 https://github.com/apache/arrow-rs/issues/4608
* No obvious way to support Attributes - https://github.com/apache/arrow-rs/issues/5435 #5334
* AsyncWrite design can easily lead to timeouts - https://github.com/apache/arrow-rs/issues/5366 https://github.com/tokio-rs/tokio/issues/4296
* The way we implement poll_flush and poll_shutdown is not entirely in keeping with the AsyncWrite contract, e.g. poll_flush may not flush all buffered data
* The ecosystem hasn't settled on a single IO trait for AsyncWrite (because they all have their own issues) - https://github.com/nrc/portable-interoperable/blob/master/io-traits/README.md
* Data is copied potentially multiple times to/from buffers
* Parallelism is controlled by the ObjectStore implementation internally with no way to control this
* AsyncWrite is tricky to integrate with synchronous code, despite the fact the internal buffering should make it straightforward
* Cannot easily track upload progress - https://github.com/apache/arrow-rs/discussions/5117

**Describe the solution you'd like**

#4971 added a `MultipartStore` abstraction that more closely mirrors the APIs exposed by object stores, avoiding all of the above issues. If we could devise a way to implement this interface for `LocalFileSystem` we could then "promote" it into the `ObjectStore` trait and deprecate put_multipart. This would provide the maximum flexibility to users, whilst being in keeping with the objectives of this crate to closely hew to the APIs of the stores themselves.

The key observation that makes this possible, is that we already recommend `MultiPartStore` be used with fixed size chunks for compatibility with r2, we therefore could require this for LocalFilesystem, in turn allowing it to support out-of-order / parallel writes as the file offsets can be determined from the part index.

https://github.com/apache/arrow-rs/pull/5431 and https://github.com/apache/arrow-rs/pull/4857 added `BufWriter` and `BufReader` and these would be retained to preserve compatibility with the tokio ecosystem and provide a more idiomatic API on top of this

**Describe alternatives you've considered**

I briefly considered a put_stream API, however, this doesn't resolve many of the above issues

We could also just implement MultipartStore for LocalFilesystem, whilst retaining the current `put_multipart`. This would allow downstreams to opt-in to the lower level API if they so wished.

We could also modify put_multipart to return something other than AsyncWrite, possibly something closer to PutPart

**Additional context**

Many of the stores also support composing objects from others, this might be something to consider in this design - https://github.com/apache/arrow-rs/issues/4966

FYI @wjones127 @Xuanwo @alamb @roeap

### LINKED ISSUE #5526: Inconsistent Multipart Nomenclature
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The object_store crate inconsistently used `multipart` `multi-part` and `multi part` to refer to the ability to upload in chunks.

**Describe the solution you'd like**

We should mirror the AWS naming, and using `multipart` as a single word - https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #5458 
Closes #5443
Closes #5526

# Rationale for this change
 

# What changes are included in this PR?

This changes put_multipart to instead return a `Box<dyn MultipartUpload`>` that provides an API much closer to that of the stores themselves. This gives users much more control over buffering and concurrency, whilst also opening up new possibilities for integrating object_store into systems such as parquet.

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/upload.rs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    ///
+    /// let (u1, u2) = futures::future::join(p1.next(), p2.next()).await;
+    /// u1.unwrap().unwrap();
+    /// u2.unwrap().unwrap();
+    /// let result = upload.complete().await.unwrap();
+    /// # }
+    /// ```
+    ///
+    /// [R2]: https://developers.cloudflare.com/r2/objects/multipart-objects/#limitations
+    fn put_part(&mut self, data: Bytes) -> UploadPart;
```
Comment (tustvold) [inline object_store/src/upload.rs:59]: The design of this is somewhat subtle.

A part index would run into issues for `LocalFilesystem` as we would need to know the offset to write the chunk to. If we permit out of order writes, e.g. writing the final partial chunk first, this chunk size would need to be provided at creation time of `Upload`. Aside from being an unfortunate API, this which would create non-trivial behaviour differences between stores that respect this config and those that ignore it.

Instead by taking a mutable borrow and providing the `data` to be written, we prohibit out of order writes. The final piece is we return a `BoxFuture<'static, Result<()>>` instead of this being an `async fn`, i.e. returning `BoxFuture<'_, Result<()>>`. This allows multiple `UploadPart` to be created and polled in parallel, without the borrow checker complaining about concurrent mutable borrows.

This is strictly a more flexible interface than the current AsyncWrite API, and whilst it still doesn't permit out of order writes, use-cases that require this can use MultiPartStore.

File: object_store/src/upload.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+use tokio::task::JoinSet;
+
+/// An upload part request
+pub type UploadPart = BoxFuture<'static, Result<()>>;
+
+#[async_trait]
+pub trait Upload: Send + std::fmt::Debug {
+    /// Upload the next part
+    ///
+    /// Returns a stream
```
Comment (tustvold) [inline object_store/src/upload.rs:31]:
```suggestion
```

File: object_store/src/upload.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    /// u2.unwrap().unwrap();
+    /// let result = upload.complete().await.unwrap();
+    /// # }
+    /// ```
+    ///
+    /// [R2]: https://developers.cloudflare.com/r2/objects/multipart-objects/#limitations
+    fn put_part(&mut self, data: Bytes) -> UploadPart;
+
+    /// Complete the multipart upload
+    async fn complete(&mut self) -> Result<PutResult>;
```
Comment (tustvold) [inline object_store/src/upload.rs:62]: Unlike `MultiPartStore` we expect `Upload` to handle the `PartId` internally, I think this makes for a more intuitive interface and avoids issues relating to the behaviour if the `PartId` are provided out of order.

File: object_store/src/lib.rs [hunk 4]
```diff
@@ -538,48 +539,11 @@ pub trait ObjectStore: std::fmt::Display + Send + Sync + Debug + 'static {
...
-    ) -> Result<(MultipartId, Box<dyn AsyncWrite + Unpin + Send>)>;
-
-    /// Cleanup an aborted upload.
-    ///
-    /// See documentation for individual stores for exact behavior, as capabilities
-    /// vary by object store.
-    async fn abort_multipart(&self, location: &Path, multipart_id: &MultipartId) -> Result<()>;
+    /// Client should prefer [`ObjectStore::put`] for small payloads, as streaming uploads
+    /// typically require multiple separate requests. See [`Upload`] for more information
+    async fn upload(&self, location: &Path) -> Result<Box<dyn Upload>>;
```
Comment (tustvold) [inline object_store/src/lib.rs:546]: Unlike before we don't expose a notion of `MultipartId`. In hindsight this makes for a confusing API. We instead encourage users to configure automatic cleanup of incomplete uploads, which is the more reliable mechanism anyway

File: object_store/src/upload.rs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+#[async_trait]
+pub trait Upload: Send + std::fmt::Debug {
+    /// Upload the next part
+    ///
+    /// Returns a stream
+    ///
+    /// Most stores require that all parts excluding the last are at least 5 MiB, and some
+    /// further require that all parts excluding the last be the same size, e.g. [R2].
+    /// Clients wanting to maximise compatibility should therefore perform writes in
+    /// fixed size blocks larger than 5 MiB.
```
Comment (crepererum) [inline object_store/src/upload.rs:36]: Could we provide an API to read these attributes from a `DynObjectStore`? Otherwise I don't see how anyone would ever be able to use `DynObjectStore` and `put_part` in combination.

File: object_store/src/upload.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            self.tasks.join_next().await.unwrap()??;
+        }
+        Ok(())
+    }
+
+    /// Write data to this [`ChunkedUpload`]
+    ///
+    /// Back pressure can optionally be applied to producers by calling
+    /// [`Self::wait_for_capacity`] prior to calling this method
+    pub fn write(&mut self, mut buf: &[u8]) {
```
[the remark this answers is not in the document]
Comment (alamb) [reply object_store/src/upload.rs:117]: It seems like this API basically requires copying the data. I wonder if there is some way to allow users to pass in an owned buffer already, like

```rust
    pub fn write(&mut self, buf: impl Into<Buffer>) {
    ...
    }
```

And then internally slicing up the Buffer to ensure the correct sizes

I realize that `put` currently requires a single contiguous buffer (per part), so maybe this copy isn't that big a problem. However it seems a pity that we require the copy 🤔 

```rust
    async fn put(&self, location: &Path, bytes: Bytes) -> Result<PutResult> {
        self.put_opts(location, bytes, PutOptions::default()).await
   }
```

File: object_store/src/upload.rs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::future::BoxFuture;
+use tokio::task::JoinSet;
+
+/// An upload part request
+pub type UploadPart = BoxFuture<'static, Result<()>>;
+
+#[async_trait]
+pub trait Upload: Send + std::fmt::Debug {
```
Comment (alamb) [inline object_store/src/upload.rs:28]: For other reviewers, This is the key new API

What is the correct description of this trait? Something like 
```suggestion
/// Represents an inprogress multi-part upload
///
/// (Is this right? Can we actually abort a multi-part upload? )
/// Cancel behavior: On drop, the multi-part upload is aborted
pub trait Upload: Send + std::fmt::Debug {
```

File: object_store/src/limit.rs [hunk 8]
```diff
@@ -81,18 +81,12 @@ impl<T: ObjectStore> ObjectStore for LimitStore<T> {
...
-    ) -> Result<(MultipartId, Box<dyn AsyncWrite + Unpin + Send>)> {
-        let permit = Arc::clone(&self.semaphore).acquire_owned().await.unwrap();
-        let (id, write) = self.inner.put_multipart(location).await?;
-        Ok((id, Box::new(PermitWrapper::new(write, permit))))
-    }
-
-    async fn abort_multipart(&self, location: &Path, multipart_id: &MultipartId) -> Result<()> {
-        let _permit = self.semaphore.acquire().await.unwrap();
-        self.inner.abort_multipart(location, multipart_id).await
+    async fn upload(&self, location: &Path) -> Result<Box<dyn Upload>> {
```
[the remark this answers is not in the document]
Comment (alamb) [reply object_store/src/limit.rs:84]: I know you prefer shorter names, but here I recommend calling this `multipart_upload` and `MultipartUpload` to align with the term used by the cloud providers. Upload is pretty generic

https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpu-upload-object.html
https://cloud.google.com/storage/docs/xml-api/post-object-complete

Or maybe keep the existing `put_multipart` name (though I realize reusing the same name may be more confusing on upgrade)

File: object_store/src/upload.rs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    /// Abort this upload
+    pub async fn abort(mut self) -> Result<()> {
+        self.tasks.shutdown().await;
+        self.upload.abort().await
+    }
+
+    /// Flush final chunk, and await completion of all in-flight requests
+    pub async fn finish(mut self) -> Result<PutResult> {
```
Comment (alamb) [inline object_store/src/upload.rs:142]: I believe this `PutResult` is what @ashtuchkin  is asking for in https://github.com/apache/arrow-rs/issues/5443

File: object_store/src/upload.rs [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+// under the License.
+
+use crate::{PutResult, Result};
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::future::BoxFuture;
+use tokio::task::JoinSet;
+
+/// An upload part request
+pub type UploadPart = BoxFuture<'static, Result<()>>;
```
[the remark this answers is not in the document]
Comment (alamb) [reply object_store/src/upload.rs:25]: Is there a `PutResult` that is returned on each part upload? As in should this be

```suggestion
pub type UploadPart = BoxFuture<'static, Result<PutResult>>;
```

**Review by alamb — COMMENTED** [review]
I think this API looks really nice @tustvold  -- thank you 🙏 

cc @wjones127 @devinjdangelo @wiedld

File: object_store/src/upload.rs (hunk 5 above)
Comment (tustvold) [reply object_store/src/upload.rs:36]: We could do, or writers could just 5 MiB chunks

File: object_store/src/upload.rs (hunk 1 above)
Comment (tustvold) [reply object_store/src/upload.rs:59]: The underlying implementations in fact all require a contiguous buffer, either for the HTTP request or to spawn the blocking IO. I did spend a decent amount of time seeing if we could avoid this, but it requires some pretty intrusive refactoring to basically extract our own Bytes abstraction. Ultimately this is still strictly better than the current API

I'll file a ticket for supporting non-contiguous payloads in ObjectStore

File: object_store/src/upload.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tustvold) [reply object_store/src/upload.rs:117]: I will file a ticket for supporting non-contiguous write payloads

**Comment by crepererum** [comment]
BTW: I really like the overall implementation!

File: object_store/src/upload.rs (hunk 1 above)
Comment (crepererum) [reply object_store/src/upload.rs:59]: > Instead by taking a mutable borrow and providing the `data` to be written, we prohibit out of order writes. The final piece is we return a `BoxFuture<'static, Result<()>>` instead of this being an `async fn`, i.e. returning `BoxFuture<'_, Result<()>>`. This allows multiple `UploadPart` to be created and polled in parallel, without the borrow checker complaining about concurrent mutable borrows.

I think this should be part of the docstring for future reference.

File: object_store/src/upload.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tustvold) [reply object_store/src/upload.rs:25]: Some APIs will return an ETag, but there isn't an obvious equivalent for LocalFilesystem, nor is it clear to me how this information would be used, and so I would rather omit it. If people need this information for whatever reason, they can use the MultipartStore

File: object_store/src/upload.rs (hunk 3 above)
Comment (tustvold) [reply object_store/src/upload.rs:62]: There isn't an obvious way this information could be used by a use of the Upload trait, but if a user needed this information they could use the lower level MultiPartStore trait

File: object_store/src/multipart.rs [hunk 11]
```diff
@@ -277,7 +43,7 @@ impl<T: PutPart> std::fmt::Debug for WriteMultiPart<T> {
 /// [`ObjectStore::put_multipart`]: crate::ObjectStore::put_multipart
 /// [`LocalFileSystem`]: crate::local::LocalFileSystem
 #[async_trait]
-pub trait MultiPartStore: Send + Sync + 'static {
+pub trait MultipartStore: Send + Sync + 'static {
```
Comment (tustvold) [inline object_store/src/multipart.rs:46]: This is a drive-by fix that has annoyed me for a while, this crate routinely uses `multipart` as a single word, so this capitalization was inconsistent

File: object_store/src/limit.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tustvold) [reply object_store/src/limit.rs]: I guess I was trying to avoid being too specific, as such a concept doesn't exist in LocalFileSystem, and Azure calls it something different

**Comment by tustvold** [comment]
I think this is now ready to go, I may refine it in subsequent PRs prior to release, but I am happy with where the core API is

File: object_store/src/upload.rs [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    /// therefore be invoked to perform this cleanup.
+    ///
+    /// It is recommended that where possible users configure lifecycle rules
+    /// to automatically reap unused parts older than some threshold, as this
+    /// will more reliably handle different failure modes. See [crate::aws] and
+    /// [crate::gcp] for more information.
+    ///
+    /// It is implementation defined behaviour to call [`MultipartUpload::abort`]
+    /// on an already completed or aborted [`MultipartUpload`]
+    async fn abort(&mut self) -> Result<()>;
```
Comment (tustvold) [inline object_store/src/upload.rs:88]: I am very tempted to just not include this, in favour of just encouraging people to configure appropriate lifecycle rules, but am curious what other people think

Comment (alamb) [reply object_store/src/upload.rs:88]: This came up downstream as well in DataFusion https://github.com/apache/arrow-datafusion/pull/9648#discussion_r1530529422

My opinion is that we should include it in the API and appropriately caveat / explain why it is not always possible. I offered a suggestion above

File: object_store/src/upload.rs [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+    /// it is implementation defined behaviour to call [`MultipartUpload::complete`]
+    /// on an already completed or aborted [`MultipartUpload`].
+    async fn complete(&mut self) -> Result<PutResult>;
+
+    /// Abort the multipart upload
+    ///
+    /// If a [`MultipartUpload`] is dropped without calling [`MultipartUpload::complete`],
+    /// some implementations will automatically reap any uploaded parts. However,
+    /// this is not always possible, e.g. for S3 and GCS. [`MultipartUpload::abort`] can
+    /// therefore be invoked to perform this cleanup.
```
Comment (alamb) [inline object_store/src/upload.rs:79]: I think the term `reap`, while more literary, has a greater potential to be misunderstood. I also think it would help to explain why it is not always possible. 

Here is a suggestion

```suggestion
    /// If a [`MultipartUpload`] is dropped without calling [`MultipartUpload::complete`],
    /// some object stores will automatically clean up any previously uploaded parts. 
    /// However, some stores such as  for S3 and GCS do not perform automatic cleanup and 
    /// in such cases, [`MultipartUpload::abort`] manually invokes this cleanup.
```

File: object_store/src/upload.rs [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    ///
+    /// It is recommended that where possible users configure lifecycle rules
+    /// to automatically reap unused parts older than some threshold, as this
+    /// will more reliably handle different failure modes. See [crate::aws] and
+    /// [crate::gcp] for more information.
```
Comment (alamb) [inline object_store/src/upload.rs:84]: I think here it helps to be explicit about "failure modes":

```suggestion
    /// Given it is not possible call `abort` in all failure scenarios (e.g. if your program is `SIGKILL`ed due to
    /// OOM), it is recommended to configure your object store with lifecycle rules
    /// to automatically cleanup unused parts older than some threshold.
    /// See [crate::aws] and [crate::gcp] for more information.
```

File: object_store/src/upload.rs [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+    /// It is implementation defined behaviour to call [`MultipartUpload::abort`]
+    /// on an already completed or aborted [`MultipartUpload`]
+    async fn abort(&mut self) -> Result<()>;
+}
+
+/// A synchronous write API for uploading data in parallel in fixed size chunks
+///
+/// Makes use of [`JoinSet`] under the hood to multiplex upload tasks,
+/// avoiding issues caused by sharing a single tokio's cooperative task
+/// budget across multiple IO operations.
```
Comment (alamb) [inline object_store/src/upload.rs:95]: I did not understand the "cooperative task budget across multiple IO operations" comment.  I think this could be simplified to something like: 

```suggestion
/// Uses multiple tokio tasks in a [`JoinSet`] to multiplex upload tasks in parallel
```

File: object_store/src/upload.rs [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+pub struct WriteMultipart {
+    upload: Box<dyn MultipartUpload>,
+
+    buffer: Vec<u8>,
+
+    tasks: JoinSet<Result<()>>,
+}
+
+impl WriteMultipart {
+    /// Create a new [`WriteMultipart`]
```
Comment (alamb) [inline object_store/src/upload.rs:112]:
```suggestion
    /// Create a new [`WriteMultipart`] that will upload using 5MB chunks
```

File: object_store/src/upload.rs [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+    pub fn new(upload: Box<dyn MultipartUpload>) -> Self {
+        Self::new_with_capacity(upload, 5 * 1024 * 1024)
```
Comment (alamb) [inline object_store/src/upload.rs:114]: nit: It might be nice to make this a named constant (mostly so you can attach a docstring to the constant and refer to it)

File: object_store/src/upload.rs [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+    pub async fn wait_for_capacity(&mut self, max_concurrency: usize) -> Result<()> {
+        while self.tasks.len() > max_concurrency {
+            self.tasks.join_next().await.unwrap()??;
+        }
+        Ok(())
+    }
+
+    /// Write data to this [`WriteMultipart`]
+    ///
+    /// Back pressure can optionally be applied to producers by calling
```
Comment (alamb) [inline object_store/src/upload.rs:136]: I think being more explicit about the implications would help here

```suggestion
    /// Note this method is synchronous (not `async`) and will immediately start new uploads
    /// as soon as the internal `capacity` is hit, regardless of 
    /// how many outstanding uploads are already in progress.
    /// 
    /// Back pressure can optionally be applied to producers by calling
```

File: object_store/src/upload.rs [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+            }
+            buf = &buf[to_read..]
+        }
+    }
+
+    fn put_part(&mut self, part: Bytes) {
+        self.tasks.spawn(self.upload.put_part(part));
+    }
+
+    /// Abort this upload
```
Comment (alamb) [inline object_store/src/upload.rs:156]:
```suggestion
    /// Abort this upload, attempting to clean up any successfully uploaded parts
```

File: object_store/src/lib.rs [hunk 20]
```diff
@@ -269,12 +269,11 @@
 //!
 //! #  Multipart Upload
 //!
-//! Use the [`ObjectStore::put_multipart`] method to atomically write a large amount of data,
-//! with implementations automatically handling parallel, chunked upload where appropriate.
+//! Use the [`ObjectStore::put_multipart`] method to atomically write a large amount of data
```
Comment (alamb) [inline object_store/src/lib.rs:272]: Can we also add an example using `BufWriter` here (to drive people to use that unless they really need to use the multi part API themselves)? It has the nice property that it does put/put-mulitpart dynamically. '

File: object_store/src/local.rs [hunk 21]
```diff
@@ -705,194 +684,91 @@ fn staged_upload_path(dest: &std::path::Path, suffix: &str) -> PathBuf {
...
-        } else if let LocalUploadState::Idle(file) = &self.inner_state {
-            let file = Arc::clone(file);
-            (&*file).write_all(buf)?;
-            Poll::Ready(Ok(buf.len()))
-        } else {
-            // If we are running on this thread, then only possible states are Idle and Complete.
-            invalid_state("when writer is already complete.")
-        }
+        let s = Arc::clone(&self.state);
+        maybe_spawn_blocking(move || {
```
Comment (alamb) [inline object_store/src/local.rs:723]: nice

File: object_store/src/buffered.rs [hunk 22]
```diff
@@ -289,9 +294,10 @@ impl AsyncWrite for BufWriter {
                         let path = std::mem::take(path);
                         let store = Arc::clone(&self.store);
                         self.state = BufWriterState::Prepare(Box::pin(async move {
-                            let (id, mut writer) = store.put_multipart(&path).await?;
-                            writer.write_all(&buffer).await?;
-                            Ok((id, writer))
+                            let upload = store.put_multipart(&path).await?;
+                            let mut chunked = WriteMultipart::new(upload);
+                            chunked.write(&buffer);
```
Comment (alamb) [inline object_store/src/buffered.rs:299]: this write results in a second buffer copy, right? That double buffering is unfortunate

**Review by alamb — APPROVED** [review]
Thank you @tustvold  TLDR is I think this PR looks really nice -- thank you for pushing it along. Overall this change looks good to me. I left several suggestions that I think are important to improve the comments, but if necessary we can do them as follow on PRs

File: object_store/src/upload.rs (hunk 15 above)
Comment (tustvold) [reply object_store/src/upload.rs:95]: I was trying to allude to issues associated with using FuturesUnordered in tokio, I'll try to dig out the ticket

**Comment by tustvold** [comment]
I intend to follow this up with a PR with some further tweaks to the BufWriter / WriteMultipart structures, but I'd like to get the changes to ObjectStore in, and then iterate on top of this
